### PR TITLE
feat(backend): Merge kfp-tekton backend code

### LINKED ISSUE #10488: [feature] Add the new Tekton packages into the KFP backends
### Feature Area
Add the new Tekton packages into the KFP backends

Most of the code are in a pluggable Tekton compiler and executor client package
https://github.com/kubeflow/kfp-tekton/tree/v2-integration/backend/src/v2/compiler/tektoncompiler
https://github.com/kubeflow/kfp-tekton/blob/v2-integration/backend/src/common/util/execution_client.go#L107-L127

### Is there a workaround currently?

---

Love this idea? Give it a 👍.

### PR DESCRIPTION
**Pull request by rimolive**
**Description of your changes:**
Closes #10488

Merge all the kfp-tekton compiler code to pipelines.

**Checklist:**
- [X] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

**Comment by rimolive** [comment]
/hold

Still working on the tests

File: backend/third_party_licenses/tekton-driver.csv [hunk 1]
```diff
@@ -0,0 +1,110 @@
+cloud.google.com/go/compute/metadata,https://github.com/googleapis/google-cloud-go/blob/compute/metadata/v0.2.3/compute/metadata/LICENSE,Apache-2.0
```
Comment (Tomcli) [inline backend/third_party_licenses/tekton-driver.csv:1]: We can remove the all 5 `tekton-*.csv`. I forgot to remove them from the kfp-tekton branch. These license was for the old implementation where we are building driver etc as a new container task.

File: backend/src/v2/objectstore/object_store.go [hunk 2]
```diff
@@ -188,7 +188,7 @@ func ParseBucketConfig(path string) (*Config, error) {
 	}
 
 	// TODO: Verify/add support for file:///.
-	if ms[1] != "gs://" && ms[1] != "s3://" && ms[1] != "minio://" {
+	if ms[1] != "gs://" && ms[1] != "s3://" && ms[1] != "minio://" && ms[1] != "mem://" {
```
Comment (Tomcli) [inline backend/src/v2/objectstore/object_store.go:191]: `mem://` to support in memory storage for launcher tests under environments that doesn't have access to public gcs endpoints

File: backend/src/agent/persistence/client/pipeline_client.go [hunk 3]
```diff
@@ -23,6 +23,7 @@ import (
 	"google.golang.org/grpc/metadata"
 
 	api "github.com/kubeflow/pipelines/backend/api/v1beta1/go_client"
+	apiv2 "github.com/kubeflow/pipelines/backend/api/v2beta1/go_client"
```
Comment (Tomcli) [inline backend/src/agent/persistence/client/pipeline_client.go:26]: Move report client to use the new v2 api

File: backend/src/agent/persistence/main.go [hunk 4]
```diff
@@ -62,6 +63,7 @@ const (
 	numWorkerName                         = "numWorker"
 	clientQPSFlagName                     = "clientQPS"
 	clientBurstFlagName                   = "clientBurst"
+	executionTypeFlagName                 = "executionType"
```
Comment (Tomcli) [inline backend/src/agent/persistence/main.go:66]: Flag to change executionType, default is `Argo`

**Comment by rimolive** [comment]
@Tomcli @chensun I need a help in here. The e2e tests that are failling are job tests, and I just ran locally but they're passing. Did I miss someting?

**Review by Tomcli — COMMENTED** [review]
Thanks @rimolive for resolving the tests. I don't have access the the backend test, but it seems like only the api integration test is failing.

**Comment by chensun** [comment]
The failure are on jobs, aka scheduled workflow:

```
integration-test-dg828-1458653792: === RUN   TestJobApi/TestJobApis
integration-test-dg828-1458653792:     job_api_test.go:374: 
integration-test-dg828-1458653792:         	Error Trace:	/go/src/github.com/kubeflow/pipelines/backend/test/integration/job_api_test.go:374
integration-test-dg828-1458653792:         	Error:      	Expected nil, but got: &errors.errorString{s:"expected runs to be length 1, got: 0"}
integration-test-dg828-1458653792:         	Test:       	TestJobApi/TestJobApis
integration-test-dg828-1458653792:     job_api_test.go:395: 
integration-test-dg828-1458653792:         	Error Trace:	/go/src/github.com/kubeflow/pipelines/backend/test/integration/job_api_test.go:395
integration-test-dg828-1458653792:         	Error:      	Expected nil, but got: &errors.errorString{s:"expected runs to be length 1, got: 0"}
integration-test-dg828-1458653792:         	Test:       	TestJobApi/TestJobApis
integration-test-dg828-1458653792: === RUN   TestJobApi/TestJobApis_SwfNotFound
integration-test-dg828-1458653792: === RUN   TestJobApi/TestJobApis_noCatchupOption
integration-test-dg828-1458653792:     job_api_test.go:512: 
integration-test-dg828-1458653792:         	Error Trace:	/go/src/github.com/kubeflow/pipelines/backend/test/integration/job_api_test.go:512
integration-test-dg828-1458653792:         	Error:      	Expected nil, but got: &errors.errorString{s:"expected runsWhenCatchupTrue with periodic schedule to be 2, got: 0"}
integration-test-dg828-1458653792:         	Test:       	TestJobApi/TestJobApis_noCatchupOption
integration-test-dg828-1458653792:     job_api_test.go:540: 
integration-test-dg828-1458653792:         	Error Trace:	/go/src/github.com/kubeflow/pipelines/backend/test/integration/job_api_test.go:540
integration-test-dg828-1458653792:         	Error:      	Expected nil, but got: &errors.errorString{s:"expected runsWhenCatchupFalse with periodic schedule to be 1, got: 0"}
integration-test-dg828-1458653792:         	Test:       	TestJobApi/TestJobApis_noCatchupOption
integration-test-dg828-1458653792: --- FAIL: TestJobApi (176.21s)
integration-test-dg828-1458653792:     --- FAIL: TestJobApi/TestJobApis (93.56s)
integration-test-dg828-1458653792:     --- PASS: TestJobApi/TestJobApis_SwfNotFound (0.50s)
integration-test-dg828-1458653792:     --- FAIL: TestJobApi/TestJobApis_noCatchupOption (81.82s)
```

Not sure how you tested this locally, consider adding debug logs to the test functions?
https://github.com/kubeflow/pipelines/blob/ac399315e66d6ed2666dc9dbaecbce4938f87356/backend/test/integration/job_api_test.go#L156C27-L156C38

File: backend/src/agent/persistence/client/pipeline_client.go (hunk 3 above)
Comment (Tomcli) [inline backend/src/agent/persistence/client/pipeline_client.go:26]: Hi @rimolive, you can revert the changes for `apiv2` because we are not planning to change the scheduledworkflow in this PR. This won't break the persistence agent because this piece of code is for sending the workflow update to the scheduled workflow service.

We can open another PR for making scheduledworkflow backend agnostic so it's easier to debug and revert in the future.

File: backend/Dockerfile.scheduledworkflow [hunk 5]
```diff
@@ -45,6 +45,7 @@ COPY --from=builder /tmp/licenses.csv /third_party/licenses.csv
 COPY --from=builder /tmp/NOTICES /third_party/NOTICES
 
 ENV NAMESPACE ""
+ENV EXECUTIONTYPE Workflow
 ENV LOG_LEVEL info
 
-CMD /bin/controller --logtostderr=true --namespace=${NAMESPACE} --logLevel=${LOG_LEVEL}
+CMD /bin/controller --logtostderr=true --namespace=${NAMESPACE} --executionType ${EXECUTIONTYPE} --logLevel=${LOG_LEVEL}
```
Comment (Tomcli) [inline backend/Dockerfile.scheduledworkflow:51]: @rimolive you can revert this change because we removed the new `executionType` flag in scheduled workflow in this PR.

**Comment by rimolive** [comment]
/unhold

@Tomcli @chensun Finally I got the tests passing but unfortunately I needed to follow Tommy's recommentation to remove the scheduled workflow and rethink the kfp-tekton code for that.

Feel free to lgtm/approve, but take into consideration that this will be a partial feature. I believe it's not a big deal given that historically we could handle partial feature implementations to make progress.

**Review by Tomcli — APPROVED** [review]
thanks @rimolive for fixing some of the tests and dependencies.
/lgtm

For the scheduledworkflow, we probably need to also update the api test to use the new swf v2 client. Let's do that in a follow up PR so it won't block the core KFP feature.

**Comment by chensun** [comment]
> /unhold
> 
> @Tomcli @chensun Finally I got the tests passing but unfortunately I needed to follow Tommy's recommentation to remove the scheduled workflow and rethink the kfp-tekton code for that.
> 
> Feel free to lgtm/approve, but take into consideration that this will be a partial feature. I believe it's not a big deal given that historically we could handle partial feature implementations to make progress.

@rimolive Just to confirm, by partial feature you meant scheduledworkflow doesn't work on kfp-tekton yet, but there's no regression to the feature on kfp-argo, right?

**Comment by Tomcli** [comment]
> > /unhold
> > @Tomcli @chensun Finally I got the tests passing but unfortunately I needed to follow Tommy's recommentation to remove the scheduled workflow and rethink the kfp-tekton code for that.
> > Feel free to lgtm/approve, but take into consideration that this will be a partial feature. I believe it's not a big deal given that historically we could handle partial feature implementations to make progress.
> 
> @rimolive Just to confirm, by partial feature you meant scheduledworkflow doesn't work on kfp-tekton yet, but there's no regression to the feature on kfp-argo, right?

Yes, there's no regression on all the kfp-argo feature.

File: backend/src/apiserver/resource/resource_manager.go [hunk 6]
```diff
@@ -517,6 +517,21 @@ func (r *ResourceManager) CreateRun(ctx context.Context, run *model.Run) (*model
...
+		job, err := r.jobStore.GetJob(run.RecurringRunId)
+		if err != nil {
+			return nil, util.NewInternalServerError(util.NewInvalidInputError("RecurringRunId doesn't exist: %s", run.RecurringRunId), "Failed to create a run due to invalid recurring run id")
+		}
+		swf, err := r.swfClient.ScheduledWorkflow(job.Namespace).Get(ctx, job.K8SName, v1.GetOptions{})
+		if err != nil {
+			return nil, util.NewInternalServerError(util.NewInvalidInputError("ScheduledWorkflow doesn't exist: %s", job.K8SName), "Failed to create a run due to invalid name")
+		}
+		executionSpec.SetOwnerReferences(swf)
+		// executionSpec.SetExecutionName(run.Description)
```
Comment (chensun) [inline backend/src/apiserver/resource/resource_manager.go:532]: nit: remove?

File: backend/src/common/util/execution_spec.go [hunk 7]
```diff
@@ -263,8 +284,49 @@ func ScheduleSpecToExecutionSpec(
 		workflow.APIVersion = "argoproj.io/v1alpha1"
 		workflow.Kind = "Workflow"
 		return NewWorkflow(workflow), nil
+	case TektonPipelineRun:
+		if executionSpecStr, ok := wfr.Spec.(string); ok {
+			return NewPipelineRunFromScheduleWorkflowSpecBytesJSON([]byte(executionSpecStr))
+		}
+		// fall back to Argo WorkflowSpec, need to marshal back to json string then unmarshal to
```
Comment (chensun) [inline backend/src/common/util/execution_spec.go:291]: nit: update comment

File: backend/src/common/util/execution_spec_test.go [hunk 8]
```diff
@@ -135,7 +135,7 @@ func TestExecutionSpec_NewExecutionSpecFromInterface(t *testing.T) {
 	execSpec, err = NewExecutionSpecFromInterface(TektonPipelineRun, test)
 	assert.Empty(t, execSpec)
 	assert.Error(t, err)
-	assert.EqualError(t, err, "InternalServerError: type:PipelineRun: ExecutionType is not supported")
+	assert.EqualError(t, err, "Invalid input error: not PipelineRun struct")
```
Comment (chensun) [inline backend/src/common/util/execution_spec_test.go:138]: nit: there should be a better error message

File: backend/src/common/util/pipelinerun.go [hunk 9]
```diff
@@ -0,0 +1,967 @@
...
+			glog.Warningf("Unrecognized input parameter: %v", k)
+		}
+	}
+	return nil
+}
+
+// Get converts this object to a workflowapi.Workflow.
+// func (pr *PipelineRun) Get() *pipelineapi.PipelineRun {
+// 	return pr.PipelineRun
+// }
```
Comment (chensun) [inline backend/src/common/util/pipelinerun.go:235]: nit: remove

File: backend/src/common/util/pipelinerun.go [hunk 10]
```diff
@@ -0,0 +1,967 @@
...
+	// // Delete/reset fields which indicate workflow is finished being persisted to the database
+	// delete(newWF.Labels, util.LabelKeyWorkflowPersistedFinalState)
+	// newWF.ObjectMeta.Labels[common.LabelKeyPhase] = string("Running")
+	// newWF.Status.Phase = "Running"
+	// newWF.Status.Message = ""
+	// newWF.Status.FinishedAt = metav1.Time{}
+	// if newWF.Spec.ActiveDeadlineSeconds != nil && *newWF.Spec.ActiveDeadlineSeconds == 0 {
+	// 	// if it was terminated, unset the deadline
+	// 	newWF.Spec.ActiveDeadlineSeconds = nil
+	// }
```
Comment (chensun) [inline backend/src/common/util/pipelinerun.go:630]: nit: remove all the dead code.

File: backend/src/common/util/pipelinerun.go [hunk 11]
```diff
@@ -0,0 +1,967 @@
...
+		// User metrics just have name and value but no NodeId.
+		metric.NodeId = taskrunStatus.PipelineTaskName
+	}
+	return reportMetricsRequest.GetMetrics(), nil
+}
+
+func readTaskRunMetricsJSONOrEmpty(
+	runID string, nodeStatus pipelineapi.PipelineRunTaskRunStatus,
+	retrieveArtifact RetrieveArtifact) (string, error) {
+	// Tekton doesn't support any artifact spec, artifact records are done by our custom metadata writers:
```
Comment (chensun) [inline backend/src/common/util/pipelinerun.go:738]: ditto: remove dead code

File: backend/Dockerfile.persistenceagent [hunk 12]
```diff
@@ -51,4 +51,4 @@ ENV TTL_SECONDS_AFTER_WORKFLOW_FINISH 86400
 ENV NUM_WORKERS 2
 ENV LOG_LEVEL info
 
-CMD persistence_agent --logtostderr=true --namespace=${NAMESPACE} --ttlSecondsAfterWorkflowFinish=${TTL_SECONDS_AFTER_WORKFLOW_FINISH} --numWorker ${NUM_WORKERS} --logLevel=${LOG_LEVEL}
```
[the remark this answers is not in the document]
Comment (rimolive) [reply backend/Dockerfile.persistenceagent:54]: Nice catch! I will rollback that change.

[the remark this answers is not in the document]
Comment (rimolive) [reply backend/Dockerfile.persistenceagent:54]: Actually, this line is completely wrong, it should have the `EXECUTIONTYPE` env var set by default and send as a flag. I'm fixing it.

File: backend/src/agent/persistence/main.go [hunk 13]
```diff
@@ -158,6 +163,7 @@ func init() {
 	// k8s.io/client-go/rest/config.go#RESTClientFor
 	flag.Float64Var(&clientQPS, clientQPSFlagName, 5, "The maximum QPS to the master from this client.")
 	flag.IntVar(&clientBurst, clientBurstFlagName, 10, "Maximum burst for throttle from this client.")
+	flag.StringVar(&executionType, executionTypeFlagName, "Workflow", "Custom Resource's name of the backend Orchestration Engine")
```
[the remark this answers is not in the document]
Comment (rimolive) [reply backend/src/agent/persistence/main.go:166]: This code uses the `kind` field of the pipeline to find its execution type. Changing to `ArgoWorkflow` would break the code.

We can think of a better way to do this, maybe using annotation. However, I think this is out of scope of this PR.

File: backend/src/agent/persistence/worker/swf_saver.go [hunk 14]
```diff
@@ -53,6 +53,10 @@ func (c *ScheduledWorkflowSaver) Save(key string, namespace string, name string,
 
 	}
 
+	// TODO: wait for up stream to officially update to v2beta1
```
[the remark this answers is not in the document]
Comment (rimolive) [reply backend/src/agent/persistence/worker/swf_saver.go:56]: But `v2beta1` is already released, right? If so, we can just remove the comment.

File: backend/src/common/util/execution_spec_test.go (hunk 8 above)
Comment (rimolive) [reply backend/src/common/util/execution_spec_test.go:138]: It follows the same message for Argo workflow: https://github.com/kubeflow/pipelines/blob/master/backend/src/common/util/workflow.go#L94

**Review by chensun — APPROVED** [review]
/lgtm
/approve

Thanks!
